packages/manager/src/features/CloudPulse/Utils/CloudPulseWidgetUtils.test.ts

```diff
@@ -0,0 +1,129 @@
+import { formatPercentage } from 'src/utilities/statMetrics';
+
+import {
+  generateGraphData,
+  generateMaxUnit,
+  getDimensionName,
+  getLabelName,
+  mapResourceIdToName,
+} from './CloudPulseWidgetUtils';
+
+import type { CloudPulseMetricsResponse } from '@linode/api-v4';
+
+it('test generateMaxUnit method', () => {
+  const legendRowsData = [
+    {
+      data: { average: 1500, last: 1600, length: 3, max: 2000, total: 4500 },
+      format: formatPercentage,
+      legendColor: '#000',
+      legendTitle: 'linode-1',
+    },
+    {
+      data: {
+        average: 2000000,
+        last: 2100000,
+        length: 3,
+        max: 2500000,
+        total: 6000000,
+      },
+      format: formatPercentage,
+      legendColor: '#000',
+      legendTitle: 'linode-2',
+    },
+  ];
+
+  const result = generateMaxUnit(legendRowsData, 'Bytes');
+  expect(result).toBe('MB');
+});
+
+it('test getLabelName method', () => {
+  const props = {
+    flags: {
+      aclpResourceTypeMap: [
+        { dimensionKey: 'resource_id', serviceType: 'linode' },
+      ],
+    },
+    label: 'CPU Usage',
+    metric: { resource_id: '123' },
+    resources: [{ id: '123', label: 'linode-1' }],
+    serviceType: 'linode',
+    unit: '%',
+  };
+
+  const result = getLabelName(props);
+  expect(result).toBe('linode-1');
```

Comment:
same here, can we think all angles and possible edge cases, if we cover that we can uncover more issues in the code and avoid UI breaking also if we are lucky